code/src/mlx90640_processor.cpp

```diff
@@ -0,0 +1,106 @@
+#include <mlx90640_processor.hpp>
+
+namespace r2d2::thermal_camera {
+
+    mlx90640_processor_c::mlx90640_processor_c(mlx90640_i2c_c &bus)
+        : bus(bus), Kvdd(0), Vdd25(0) {
+    }
+
+    void mlx90640_processor_c::set_and_read_raw_pixels() {
+        // TODO: To be implemented yet
+    }
+
+    uint8_t mlx90640_processor_c::get_resolution_correlation() const {
+        uint16_t resolution_ee = extract_data(EE_RESOLUTION, 0x3000, 12);
+        uint16_t resolution_reg =
+            extract_data(INTERNAL_CONTROL_REGISTER, 0xFFFF, 10);
+        resolution_ee = std::pow(2, resolution_ee);
+        resolution_reg = std::pow(2, resolution_reg);
+        return static_cast<uint8_t>(resolution_ee / resolution_reg);
+    }
+
+    uint16_t mlx90640_processor_c::extract_data(const uint16_t reg_addr,
+                                                const uint16_t and_bits,
+                                                const uint8_t shifted) const {
+        return (bus.read_register(reg_addr) & and_bits) >> shifted;
+    }
+
+    void mlx90640_processor_c::apply_treshold(int &value,
+                                              const uint16_t exceeds,
+                                              const int minus) const {
+        if (value > exceeds) {
+            value -= minus;
+        }
+    }
+
+    int mlx90640_processor_c::get_compensated_data(const uint16_t reg_addr,
+                                                   const uint16_t and_bits,
+                                                   const uint8_t shifted,
+                                                   const uint16_t exceeds,
+                                                   const int minus) const {
+        int data = extract_data(reg_addr, and_bits, shifted);
+        apply_treshold(data, exceeds, minus);
+        return data;
+    }
+
+    float mlx90640_processor_c::get_Vdd() {
+        Kvdd = get_compensated_data(EE_VDD_PIX, 0xFF00, 8, 127, 256);
+        Kvdd *= 32;
+
+        Vdd25 = extract_data(EE_VDD_PIX, 0x00FF, 0);
+        Vdd25 = (Vdd25 - 256) * 32 - 8192;
+
+        int ram_vdd_pix = bus.read_register(RAM_VDD_PIX);
+        apply_treshold(ram_vdd_pix, 32767, 65536);
+
+        uint8_t res_cor = get_resolution_correlation();
+        float Vdd = ((res_cor * ram_vdd_pix - Vdd25) / Kvdd) + VDD0;
+        return Vdd;
+    }
+
+    float mlx90640_processor_c::get_Ta() const {
```

Comment:
A lot of things in this function look like magic number to me, an outside reviewer, please specify where these numbers come from.

Reply:
The datasheet does not come with clear explanations about what numbers are.

Reply:
Please refer to the datasheet pages in a comment so people at least know hoe you got the numbers

Reply:
You might also want to link the datasheet in the code since datasheets can differ from each other even when it's about the same chip

Reply:
@JasperOosterbroek this is fine by me, datasheet should be a resource on the wiki.